### ISSUE TITLE
Comprehension type inference error

### ISSUE THREAD

**Issue by ehsantn** [numba/numba#2597]
The following example fails in the stencil branch but not master:

```python
import numba
import numpy as np

@numba.njit(parallel=True)
def comp_test(A, s):
    N, D = A.shape
    V = np.array([5.0 for i in range(N)])
    B = np.array([[3.0 for j in range(D)] for i in range(s)])
    return B[0,0]

A = np.random.ranf((100, 4))
print(comp_test(A, 5))
```
```
  File "/Users/etotoni/pse-hpc/python/numba/numba/compiler.py", line 452, in stage_nopython_frontend
    self.locals)
  File "/Users/etotoni/pse-hpc/python/numba/numba/compiler.py", line 865, in type_inference_stage
    infer.propagate()
  File "/Users/etotoni/pse-hpc/python/numba/numba/typeinfer.py", line 844, in propagate
    raise errors[0]
  File "/Users/etotoni/pse-hpc/python/numba/numba/typeinfer.py", line 137, in propagate
    constraint(typeinfer)
  File "/Users/etotoni/pse-hpc/python/numba/numba/typeinfer.py", line 559, in __call__
    (targetty, idxty, valty), loc=self.loc)
numba.errors.TypingError: Failed at nopython (nopython frontend)
Cannot resolve setitem: array(undefined, 1d, C)[int64] = array(float64, 1d, C)
```

**Comment by ehsantn** [comment]
Ping @ninegua @sklam @DrTodd13

**Comment by ninegua** [comment]
Ok, should be fixed by 84ffa291a10f89e36aa8032cb5c6eb383c070908

**Comment by ehsantn** [comment]
Thanks!
